### ISSUE TITLE
Discussion about "DeepInspect: A Black-box Trojan Detection and Mitigation Framework for Deep Neural Networks"

### ISSUE THREAD

**Issue by tgy97** [THUYimingLi/backdoor-learning-resources#13]
DeepInspect is claimed as a "black-box" detection method, that is, "does not need any information about model internals" as the paper said.

However, it seems that the Loss_trigger used in eq.3 is based on the propagation through the model to be tested.
![捕获](https://user-images.githubusercontent.com/33145739/114124634-76276680-9927-11eb-8efd-28d2c43d9ede.PNG)

I was confused about how to perform backpropagation without the access to model parameter. Or, did I misunderstand the method in this paper?

**Comment by THUYimingLi** [comment]
I go through its method and it seems that equation (3) is used to update the GAN, while the model D is fixed. In other words, the author needs to calculate the L_trigger w.r.t. GAN' parameters rather than D's parameters. It can be calculated given the fact that the defender has the following knowledge about the inquired DNN: dimensionality of the input data, number of output classes, and the confidence scores of the model given an arbitrary input query.

Moreover, I guess it would be better to send an email to authors of this paper. I guess they can provide more details and insights.

**Comment by THUYimingLi** [comment]
> DeepInspect is claimed as a "black-box" detection method, that is, "does not need any information about model internals" as the paper said.
> 
> However, it seems that the Loss_trigger used in eq.3 is based on the propagation through the model to be tested.
> ![捕获](https://user-images.githubusercontent.com/33145739/114124634-76276680-9927-11eb-8efd-28d2c43d9ede.PNG)
> 
> I was confused about how to perform backpropagation without the access to model parameter. Or, did I misunderstand the method in this paper?

I go through its method and it seems that equation (3) is used to update the GAN, while the model D is fixed. In other words, the author needs to calculate the L_trigger w.r.t. GAN' parameters rather than D's parameters. It can be calculated given the fact that the defender has the following knowledge about the inquired DNN: dimensionality of the input data, number of output classes, and the confidence scores of the model given an arbitrary input query.

Moreover, I guess it would be better to send an email to authors of this paper. I guess they can provide more details and insights.

**Comment by tgy97** [comment]
I see. Eq.3 is indeed used for updating the GAN, not the model D. But from the equation, the output of GAN is sequentially passed to the model D, and the L_trigger is caculated with the output of D. So the backpropagation path still contains the internal value of model D, which seems more like a white-box setting. 

Of course, there are some score-based black-box adversarial attack methods like ZOO and NES, which could be used to calculate the L_trigger w.r.t GAN' parameter with only the confidence scores given by the model D. I'm not sure whether the authors used a similar method.

I'll try to contact to the author and update this issue if I get a response.

**Comment by tgy97** [comment]
I have communicated with the author through several emails. The author explains that the internal information of the model D is not needed to optimize the generator G. 

In my last letter to the author, I ask how to optimize the generator G without the internal value of D and I write a small piece of pseudo code to describe the method from my understanding.
![image](https://user-images.githubusercontent.com/33145739/114497495-ddf5ee00-9c54-11eb-84da-601294cc34e3.png)

The author replys that the pseudo code makes sense.

But in my opinion, the method described in the pseudo code is totally a white-box procedure, where the loss can back-propagate through the model D directly.

Do I misunderstand the black-box setting? How can loss back-propagate through the model directly even without the internal information of the model?

**Comment by THUYimingLi** [comment]
> I have communicated with the author through several emails. The author explains that the internal information of the model D is not needed to optimize the generator G.
> 
> In my last letter to the author, I ask how to optimize the generator G without the internal value of D and I write a small piece of pseudo code to describe the method from my understanding.
> ![image](https://user-images.githubusercontent.com/33145739/114497495-ddf5ee00-9c54-11eb-84da-601294cc34e3.png)
> 
> The author replys that the pseudo code makes sense.
> 
> But in my opinion, the method described in the pseudo code is totally a white-box procedure, where the loss can back-propagate through the model D directly.
> 
> Do I misunderstand the black-box setting? How can loss back-propagate through the model directly even without the internal information of the model?

Yes, I think you are correct. It is impossible to update the GAN without obtaining the gradient of the model D, unless they adopted some zero-order optimization methods (such as those adopted in black-box adversarial attack, as you mentioned).

**Comment by THUYimingLi** [comment]
BTW, there is at least one recent work, which is in the black-box settings. You can found it here (https://arxiv.org/pdf/2103.13127.pdf). Moreover, I will close this issue if you do not have any further question :)

**Comment by tgy97** [comment]
Thanks. I think this issue can be closed.
